components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt

```diff
@@ -0,0 +1,150 @@
+                return@let it.substring(0, MAX_ID_LENGTH)
+            }
+            it
+        }
+
+        val experimentData = RecordedExperimentData(truncatedBranch, extra)
+        experiments.put(truncatedExperimentId, experimentData)
+    }
+
+    /**
+     * Set an active experiment as inactive.
+     *
+     * Warns to the logger (but does not raise) if the experiment is not already active.
+     *
+     * @param experimentId the id of the experiment to set as inactive
+     */
+    fun setExperimentInactive(experimentId: String) {
+        val truncatedExperimentId = experimentId.let {
+            if (it.length > MAX_ID_LENGTH) {
+                logger.warn("experimentId ${it.length} > $MAX_ID_LENGTH")
+                return@let it.substring(0, MAX_ID_LENGTH)
+            }
+            it
+        }
+
+        if (!experiments.containsKey(truncatedExperimentId)) {
+            logger.warn("experiment $experimentId} not already active")
+        } else {
+            experiments.remove(truncatedExperimentId)
+        }
+    }
+
+    /**
+     * Retrieves the data about active experiments
+     *
+     * @return the map of experiment ids to [RecordedExperimentData]
+     */
+    @Synchronized
+    fun getSnapshot(): Map<String, RecordedExperimentData> {
+        return experiments
+    }
+
+    /**
+     * A helper function to get the extra data as JSON.
+     */
+    private fun getExtraAsJSON(extra: Map<String, String>): JSONObject {
+        val jsonExtra = JSONObject()
+        for ((k, v) in extra) {
+            jsonExtra.put(k, v)
+        }
+        return jsonExtra
+    }
+
+    /**
+     * Get a snapshot of the stored data as a JSON object.
+     *
+     * @param storeName the name of the desired store.  Since experiments are
+     *   included in all stores, this parameter is effectively ignored, but
+     *   must be provided to comply with the StorageEngine interface.
+     * @param clearStore whether or not to clearStore the requested store.
+     *   This parameter is ignored.
+     *
+     * @return the JSONObject containing information about the active experiments
+     */
+    override fun getSnapshotAsJSON(
+        @Suppress("UNUSED_PARAMETER") storeName: String,
+        @Suppress("UNUSED_PARAMETER") clearStore: Boolean
+    ): Any? {
+        return getSnapshot().let { pingExperiments ->
+            if (pingExperiments.count() == 0) {
+                return null
+            }
+
+            val experimentsMap = JSONObject()
+            for ((key, value) in pingExperiments) {
+                val experimentData = JSONArray()
+                experimentData.put(value.branch)
+                if (value.extra != null) {
+                    experimentData.put(getExtraAsJSON(value.extra))
+                }
+                experimentsMap.put(key, experimentData)
+            }
+            return experimentsMap
+        }
+    }
+
+    override val sendAsTopLevelField: Boolean
+        get() = true
```

Comment:
nit: If this is always true you can also create a backing field for it and set it once:

```Kotlin
override val sendAsTopLevelField: Boolean = true
```